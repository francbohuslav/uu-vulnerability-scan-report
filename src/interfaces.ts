export interface ICommandOptions {
  /**
   * File with report from sonatype
   */
  sourceReport: string;
  outputFile: string;
  /**
   * Report is in TeamCity format
   */
  teamcity: boolean;
  /**
   * only: only UU libraries are verified, except: verify all libraries except UU libraries
   */
  uu: WhiteListTypeEnum;
}

export interface INode {
  library: string;
  version: string;
  level: number;
  vulnerabilities: IVulnerability[];
  childNodes: INode[];
}
export interface IVulnerability {
  cve?: string;
  cwe?: string;
  title: string;
  importance: string;
  url: string;
}

export enum WhiteListTypeEnum {
  all = "",
  onlyUu = "only",
  exceptUu = "except",
}

export type IVulnerabilities = { [vulnerabilityUrlOrCve: string]: IVulnerableOccurrence };
export interface IVulnerableOccurrence {
  vulnerability: IVulnerability;
  usedBy: ILibraryPath[];
}

export type ILibraryPath = string[];

export interface IConsole {
  log(...args: string[]): void;
  error(...args: string[]): void;
}
