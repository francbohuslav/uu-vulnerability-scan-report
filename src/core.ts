const prompt = require("prompt-sync")();
import { spawn } from "child_process";
import fs from "fs";

class Core {
    private locationHistory: string[] = [];
    constructor() {}

    /**
     * Ask for user input
     * @param {string} question
     * @param {string} defaultValue
     * @return {string}
     */
    public ask(question: string, defaultValue: string = null): string | boolean {
        if (defaultValue === null) {
            question = question + " [Y/n]";
        } else {
            defaultValue = defaultValue.toUpperCase();
            question = question + " Default=" + defaultValue;
        }
        let answer = prompt(question + " ");
        if (answer === null) {
            // CTRL + C
            this.showError("Terminated by user");
        }
        answer = answer.toUpperCase();
        if (!defaultValue) {
            return answer === "" || answer === "Y";
        } else {
            answer = answer.toUpperCase();
            if (answer === "") {
                answer = defaultValue;
            }
            return answer;
        }
    }

    public showMessage(message: string): void {
        console.log("\x1b[36m%s\x1b[0m", message);
        // Title
        process.stdout.write(String.fromCharCode(27) + "]0;" + message + String.fromCharCode(7));
    }

    public showSuccess(message: string): void {
        console.log("\x1b[32m%s\x1b[0m", message);
    }

    public showWarning(message: string): void {
        console.log("\x1b[33m%s\x1b[0m", message);
    }

    public showError(message: string, exit = true): void {
        console.log("\x1b[31m%s\x1b[0m", message);
        //rompt("press ENTER to continue ");
        if (exit) {
            process.exit(1);
        }
    }

    public async delay(ms: number): Promise<void> {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve(null);
            }, ms);
        });
    }

    /**
     *
     * @param {string} command
     * @param {string[]} args
     */
    public async runCommand(command: string, args?: string[] | string, options?: { disableStdOut?: boolean }): Promise<ICommandResponse> {
        if (!args || args.length === 0) {
            if (command.indexOf(" ") > -1) {
                args = command.split(" ");
                command = args.shift();
            }
        } else {
            if (typeof args === "string") {
                args = args.split(" ");
            }
        }
        return new Promise((resolve, reject) => {
            let stdOut = "";
            let stdErr = "";
            const context = spawn(command, args as string[]);
            context.stdout.on("data", (data) => {
                stdOut += data.toString();
                if (!options || !options.disableStdOut) {
                    console.log(data.toString());
                }
            });

            context.stderr.on("data", (data) => {
                stdErr += data.toString();
                if (data.toString().toLowerCase().indexOf("warn") > -1) {
                    console.log("\x1b[33m%s\x1b[0m", data.toString());
                } else {
                    console.log("\x1b[31m%s\x1b[0m", data.toString());
                }
            });

            context.on("error", (error) => {
                console.log("\x1b[31m%s\x1b[0m", error.message);
            });

            context.on("close", (code) => {
                if (code) {
                    reject({ code, stdOut, stdErr } as ICommandResponse);
                } else {
                    resolve({ stdOut, stdErr } as ICommandResponse);
                }
            });
        });
    }

    public runCommandNoWait(command: string, args: string[] | string) {
        if (!args || args.length === 0) {
            if (command.indexOf(" ") > -1) {
                args = command.split(" ");
                command = args.shift();
            }
        } else {
            if (typeof args === "string") {
                args = args.split(" ");
            }
        }
        spawn(command, args as string[], {
            detached: true,
            shell: true,
            stdio: "ignore",
        });
    }

    public processExit(errorLevel: string, location: string) {
        if (errorLevel) {
            this.showError(`ERROR: ${errorLevel}` + (location ? " in " + location : ""));
        }
    }

    public async inLocationAsync(path: string, asyncAction: () => Promise<void>) {
        this.pushLocation(path);
        await asyncAction();
        this.popLocation();
    }

    /**
     * @param {string} path
     * @param {function} action
     */
    public inLocation(path: string, action: () => void) {
        this.pushLocation(path);
        action();
        this.popLocation();
    }

    public pushLocation(path: string) {
        const wd = process.cwd();
        this.locationHistory.push(wd);
        process.chdir(path);
    }

    public popLocation() {
        if (this.locationHistory.length) {
            const wd = this.locationHistory.pop();
            process.chdir(wd);
            return wd;
        } else {
            this.showError("There is no location to pop");
        }
    }

    /**
     *
     * @param {string} file
     */
    public readTextFile(file: string) {
        return fs.readFileSync(file, { encoding: "utf-8" }).toString();
    }

    public writeTextFile(filePath: string, data: string) {
        fs.writeFileSync(filePath, data, { encoding: "utf-8" });
    }

    public async getProcessIdByPort(port: string) {
        const data = await this.runCommand(`netstat -ano`, undefined, { disableStdOut: true });
        const lines = data.stdOut.split(/[\r\n]+/);
        const portLines = lines.filter((line) => line.indexOf("0.0.0.0:" + port) > -1);
        const match = portLines[0] && portLines[0].match(/\d+$/);
        if (match) {
            return match[0];
        }
        return false;
    }
}
export default new Core();

interface ICommandResponse {
    code?: number;
    stdOut: string;
    stdErr: string;
}

/*
Colors reference

Reset = "\x1b[0m"
Bright = "\x1b[1m"
Dim = "\x1b[2m"
Underscore = "\x1b[4m"
Blink = "\x1b[5m"
Reverse = "\x1b[7m"
Hidden = "\x1b[8m"

FgBlack = "\x1b[30m"
FgRed = "\x1b[31m"
FgGreen = "\x1b[32m"
FgYellow = "\x1b[33m"
FgBlue = "\x1b[34m"
FgMagenta = "\x1b[35m"
FgCyan = "\x1b[36m"
FgWhite = "\x1b[37m"

BgBlack = "\x1b[40m"
BgRed = "\x1b[41m"
BgGreen = "\x1b[42m"
BgYellow = "\x1b[43m"
BgBlue = "\x1b[44m"
BgMagenta = "\x1b[45m"
BgCyan = "\x1b[46m"
BgWhite = "\x1b[47m"
*/
