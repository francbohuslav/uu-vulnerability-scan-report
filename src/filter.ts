import { INode } from "./interfaces";

export class Filter {
  public filter(nodes: INode[], include: RegExp[], exclude: RegExp[]): INode[] {
    const matchedNodes: INode[] = [];
    if (include.length === 0) {
      // Add all nodes to results
      for (const node of nodes) {
        matchedNodes.push(node);
      }
    } else {
      // Add only matched nodes
      this.searchForLibraries(nodes, matchedNodes, include);
    }
    // Remove excludes from result
    const resultNodes: INode[] = [];
    for (const node of matchedNodes) {
      if (!this.matchPattern(node, exclude)) {
        resultNodes.push(node);
        node.childNodes = this.filter(node.childNodes, [], exclude);
      }
    }
    return resultNodes;
  }

  private searchForLibraries(nodes: INode[], newNodes: INode[], includeRegExpList: RegExp[]) {
    for (const node of nodes) {
      if (this.matchPattern(node, includeRegExpList)) {
        newNodes.push(node);
      } else if (node.childNodes.length) {
        this.searchForLibraries(node.childNodes, newNodes, includeRegExpList);
      }
    }
  }

  private matchPattern(node: INode, includeRegExpList: RegExp[]): boolean {
    for (const include of includeRegExpList) {
      if (node.library.match(include)) {
        return true;
      }
    }
    return false;
  }
}
