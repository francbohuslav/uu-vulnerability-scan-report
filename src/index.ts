/* istanbul ignore file */
import commandLineArgs, { OptionDefinition } from "command-line-args";
import { ICommandOptions, WhiteListTypeEnum } from "./interfaces";
import { VulnerabilityScan } from "./vulnerabilityScan";

// See ICommandOptions for help
const optionDefinitions: OptionDefinition[] = [
  { name: "sourceReport", alias: "s", type: String },
  { name: "outputFile", alias: "o", type: String },
  { name: "teamcity", alias: "t", type: Boolean },
  /**
   * {@link WhiteListTypeEnum}
   */
  { name: "uu", alias: "u", type: String },
];
const options = commandLineArgs(optionDefinitions) as ICommandOptions;
options.uu = options.uu ?? WhiteListTypeEnum.all;
new VulnerabilityScan(options, console).run();
