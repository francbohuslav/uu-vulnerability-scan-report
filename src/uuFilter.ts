import { INode, WhiteListTypeEnum } from "./interfaces";

export class UuFilter {
  public filter(nodes: INode[], type: WhiteListTypeEnum): INode[] {
    if (type === WhiteListTypeEnum.all) {
      return nodes;
    } else if (type === WhiteListTypeEnum.exceptUu) {
      const newNodes: INode[] = [];
      for (const node of nodes) {
        if (!this.isUuDependency(node)) {
          newNodes.push(node);
          node.childNodes = this.filter(node.childNodes, type);
        }
      }
      return newNodes;
    } else {
      const newNodes: INode[] = [];
      this.searchForUuLibraries(nodes, newNodes);
      return newNodes;
    }
  }

  private searchForUuLibraries(nodes: INode[], newNodes: INode[]) {
    for (const node of nodes) {
      if (this.isUuDependency(node)) {
        newNodes.push(node);
      } else if (node.childNodes.length) {
        this.searchForUuLibraries(node.childNodes, newNodes);
      }
    }
  }

  private isUuDependency(node: INode): boolean {
    return node.library.startsWith("uu.appg01:");
  }
}
