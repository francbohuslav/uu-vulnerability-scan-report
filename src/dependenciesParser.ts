import core from "./core";
import { IConsole, INode } from "./interfaces";

export class DependenciesParser {
  constructor(private console: IConsole) {}

  public parse(file: string): INode[] {
    const content = core.readTextFile(file);
    const confs = this.getConfigurationSegments(content);
    const allNodes: INode[] = [];
    for (const conf of confs) {
      const root = this.getTree(conf);
      if (root.library === "checkstyle") {
        continue;
      }
      for (const node of root.childNodes) {
        allNodes.push(node);
      }
    }
    // core.writeTextFile("test.json", JSON.stringify(treeList, null, 2));
    return allNodes;
  }

  /**
   * Returns only lines that contains dependencies and vulnerabilites.
   * @param content
   */
  private getConfigurationSegments(content: string): IConfigurationLines[] {
    let lines = content.split(/\r?\n/);
    // Remove lock file statements
    lines = lines.filter((l) => !l.match(/\{strictly.* \(c\)$/));
    const configurations: IConfigurationLines[] = [];
    let current: IConfigurationLines = {
      configurationName: "",
      lines: [],
    };
    let fromLine = 0;
    for (let lineIndex = 0; lineIndex < lines.length; lineIndex++) {
      const line = lines[lineIndex];
      const match = line.match(/^([a-z]+)(\s-\s.*)?$/i);
      if (!current.configurationName && match) {
        current.configurationName = match[1];
        fromLine = lineIndex + 1;
      }
      if (line.trim() === "" && fromLine) {
        current.lines = lines.slice(fromLine, lineIndex);
        if (current.lines.length > 0 && current.lines[0].trim() != "No dependencies") {
          configurations.push(current);
        }
        current = { configurationName: "", lines: [] };
        fromLine = 0;
      }
    }
    return configurations;
  }

  private getTree(segment: IConfigurationLines): INode {
    const root: INode = {
      level: 0,
      library: segment.configurationName,
      childNodes: [],
      version: "",
      vulnerabilities: [],
    };
    this.createSubNode(root, root, segment.lines);
    return root;
  }

  private createSubNode(parentNode: INode, prevNode: INode, lines: string[]): INode | undefined {
    let line = lines.shift();
    while (line) {
      const idIndex = line.indexOf("---");
      const libraryText = line.substring(idIndex + 4).trim();
      // 0 => 1, 6 => 2, 11 => 3, 16 => 4
      const level = idIndex == 0 ? 1 : Math.floor((idIndex + 4) / 5);
      const match = libraryText.trim().match(/^(.+?:.+?):(.+?)$/);
      let node: INode = {
        level,
        library: match ? match[1] : libraryText,
        version: match ? this.parseVersion(match[2]) : "",
        childNodes: [],
        vulnerabilities: [],
      };

      if (prevNode.level === level) {
        parentNode.childNodes.push(node);
      } else if (level > prevNode.level) {
        prevNode.childNodes.push(node);
        const nextNode = this.createSubNode(prevNode, node, lines);
        if (nextNode?.level === parentNode.level + 1) {
          parentNode.childNodes.push(nextNode);
          node = nextNode;
        } else {
          return nextNode;
        }
      } else {
        return node;
      }
      prevNode = node;

      line = lines.shift();
    }
    return undefined;
  }

  private parseVersion(versionString: string): string {
    versionString = versionString.replace(/ \(.\)$/, "");
    const match = versionString.match(/^(.+ -> )?(\S+)$/);
    if (match) {
      return match[2];
    }
    const matchProject = versionString.match(/^(.+) -> project .*$/);
    if (matchProject) {
      return matchProject[1];
    }
    this.console.error(`Version '${versionString}' is not in pattern '2.14.1' or '2.14.1 -> 2.13.5 (c)'`);
    return versionString;
  }
}

interface IConfigurationLines {
  configurationName: string;
  lines: string[];
}
