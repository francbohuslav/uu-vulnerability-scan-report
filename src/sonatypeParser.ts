import core from "./core";
import { ICommandOptions, IConsole, INode, IParser, IVulnerability } from "./interfaces";

export class SonatypeParser implements IParser {
  private readonly libraryIdentifier = "+---";

  constructor(private console: IConsole) {}

  public parse(options: ICommandOptions): INode[] {
    const content = core.readTextFile(options.sourceReport);
    const lines = this.getTreeLines(content);
    const tree = this.getTree(lines);
    // core.writeTextFile("../testingOutput.txt", this.testingPrint(tree));
    return tree;
  }

  /* istanbul ignore next */
  testingPrint(nodes: INode[]): string {
    let str: string = "";
    for (const node of nodes) {
      str += `${"|    ".repeat(node.level - 1)}${this.libraryIdentifier} ${node.library}:${node.version}: ${
        node.vulnerabilities.length
      } vulnerabilities detected\r\n`;
      for (const vul of node.vulnerabilities) {
        if (vul.cve) {
          str += `${"|    ".repeat(Math.max(node.level - 1, 0))}     \x1b[31m[${vul.cve}] ${vul.cwe}: ${vul.title} (${vul.importance}): ${vul.url}\x1b[0m\r\n`;
        } else {
          str += `${"|    ".repeat(Math.max(node.level - 1, 0))}     \x1b[31m${vul.title} (${vul.importance}): ${vul.url}\x1b[0m\r\n`;
        }
      }
      if (node.childNodes.length > 0) {
        str += this.testingPrint(node.childNodes);
      }
    }
    return str;
  }

  private getTree(lines: string[]): INode[] {
    const root: INode = {
      level: 0,
      library: "ROOT",
      childNodes: [],
      version: "",
      vulnerabilities: [],
    };
    this.createSubNode(root, root, lines);
    return root.childNodes;
  }

  private createSubNode(parentNode: INode, prevNode: INode, lines: string[]): INode | undefined {
    let line = lines.shift();
    while (line) {
      const idIndex = line.indexOf(this.libraryIdentifier);
      if (idIndex > -1) {
        const beforeIdentifier = line.substring(0, idIndex);
        const level = beforeIdentifier.split("|").length;
        const match = line
          .substring(idIndex + this.libraryIdentifier.length)
          .trim()
          .match(/^(.*):([^:]+):\s*([^:]+)$/);
        let node: INode = {
          level,
          library: match ? match[1] : line,
          version: match ? match[2].replace(" (*)", "") : "",
          childNodes: [],
          vulnerabilities: [],
        };

        if (prevNode.level === level) {
          parentNode.childNodes.push(node);
        } else if (level > prevNode.level) {
          prevNode.childNodes.push(node);
          const nextNode = this.createSubNode(prevNode, node, lines);
          if (nextNode?.level === parentNode.level + 1) {
            parentNode.childNodes.push(nextNode);
            node = nextNode;
          } else {
            return nextNode;
          }
        } else {
          return node;
        }
        prevNode = node;
      } else {
        const vul = this.parseVulnerability(line);
        if (vul) {
          prevNode.vulnerabilities.push(vul);
        }
      }
      line = lines.shift();
    }
    return undefined;
  }

  private parseVulnerability(line: string): IVulnerability | undefined {
    // |    |         \u001b[31m[CVE-2020-36518] CWE-787: Out-of-bounds Write (7.5/10, High): https://ossindex.sonatype.org/vulnerability/CVE-2020-36518?...\u001b[0m
    line = line.replace(/^(\|\s+)*/g, "").trim();
    line = line.replace(/\u001b.*?m/g, "");
    let match = line.match(/^\[(.*)\] (.*): (.*) \((.*)\): (.*)$/);
    if (match) {
      return {
        cve: match[1],
        cwe: match[2],
        title: match[3],
        importance: match[4],
        url: match[5],
      };
    }
    // 1 vulnerability found (7.5/10, High): https://ossindex.sonatype.org/vulnerability/sonatype-2021-4682
    match = line.match(/^(.*) \((.*)\): (.*)$/);
    if (match) {
      return {
        title: match[1],
        importance: match[2],
        url: match[3],
      };
    }
    this.console.log(line);

    this.console.error(`Cannot parse: ${line}`);
  }

  /**
   * Returns only lines that contains dependencies and vulnerabilites.
   * @param content
   */
  private getTreeLines(content: string): string[] {
    let lines = content.split(/\r?\n/);
    const range = { fromLine: -1, toLine: 0 };
    for (let lineNumber = 1; lineNumber <= lines.length; lineNumber++) {
      const line = lines[lineNumber - 1];
      if (line.match(/^[+|]/)) {
        if (range.fromLine === -1) {
          range.fromLine = lineNumber;
        }
        range.toLine = lineNumber;
      }
      if (!line.includes(this.libraryIdentifier)) {
      }
    }
    return lines.slice(range.fromLine - 1, range.toLine);
  }
}
