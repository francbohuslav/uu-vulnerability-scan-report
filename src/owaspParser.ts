import core from "./core";
import { DependenciesParser } from "./dependenciesParser";
import { ICommandOptions, IConsole, INode, IParser, IVulnerability } from "./interfaces";

export class OwaspParser implements IParser {
  constructor(private console: IConsole) {}

  public parse(options: ICommandOptions): INode[] {
    const treeList = new DependenciesParser(this.console).parse(options.dependencies ?? "");
    const report = JSON.parse(core.readTextFile(options.sourceReport)) as IDCR$Root;
    for (const dependency of report.dependencies) {
      if (dependency.vulnerabilities && dependency.vulnerabilities.length > 0) {
        const vulnerabilitiesPerLibrary = this.createVulnerabilities(dependency);
        this.setVulnerabilityToNodes(treeList, vulnerabilitiesPerLibrary);
        // Can't be used because checkStyle configuration is ommited
        // this.checkAllVulnerabilitiesUsage(vulnerabilitiesPerLibrary);
      }
    }
    // core.writeTextFile("../test.json", JSON.stringify(treeList, null, 2));
    return treeList;
  }

  private createVulnerabilities(dependency: IDCR$Dependency): Map<string, IPackageVulnerabilities> {
    const vulPerPackage = new Map<string, IPackageVulnerabilities>();
    for (const dcrPackage of dependency.packages) {
      const vulnerabilities: IVulnerability[] = [];
      for (const dcrVulnerability of dependency.vulnerabilities) {
        const vulnerability: IVulnerability = {
          importance: dcrVulnerability.severity,
          title: dcrVulnerability.description?.trim(),
          url: dcrPackage.url,
          cve: dcrVulnerability.name,
          cwe: dcrVulnerability.cwes.join(", "),
        };
        vulnerabilities.push(vulnerability);
      }
      const match = dcrPackage.id.match(/^pkg:maven\/([^/]+)\/([^/]+@.+)$/);
      if (match) {
        vulPerPackage.set(`${match[1]}:${match[2]}`, { vulnerabilities, used: 0 });
      } else {
        this.console.error(`Package id '${dcrPackage.id}' does not match pattern pkg:maven/com.rabbitmq/amqp-client@5.13.1"`);
      }
    }
    return vulPerPackage;
  }

  private setVulnerabilityToNodes(treeList: INode[], vulnerabilitiesPerLibrary: Map<string, IPackageVulnerabilities>) {
    for (const tree of treeList) {
      const vuls = vulnerabilitiesPerLibrary.get(tree.library + "@" + tree.version);
      if (vuls) {
        tree.vulnerabilities = vuls.vulnerabilities;
        vuls.used++;
      }
      if (tree.childNodes.length > 0) {
        this.setVulnerabilityToNodes(tree.childNodes, vulnerabilitiesPerLibrary);
      }
    }
  }

  private checkAllVulnerabilitiesUsage(vulnerabilitiesPerLibrary: Map<string, IPackageVulnerabilities>) {
    for (const [pack, vul] of vulnerabilitiesPerLibrary.entries()) {
      if (vul.used === 0) {
        this.console.error(`Vulnerable package '${pack}' not found in dependency tree. This should never happen.`);
      }
    }
  }
}

interface IDCR$Root {
  dependencies: IDCR$Dependency[];
}

interface IDCR$Dependency {
  /** "The RabbitMQ Java client library allows Java applications to interface with RabbitMQ." */
  description: string;
  vulnerabilities: IDCR$Vulnerability[];
  packages: IDCR$Package[];
}

interface IDCR$Vulnerability {
  /** "CVE-2023-46120" */
  name: string;

  /** "NVD" */
  source: string;

  /** "HIGH" */
  severity: string;

  /** [ "CWE-400" ] */
  cwes: string[];
  description: string;
}

interface IDCR$Package {
  /** "pkg:maven/com.rabbitmq/amqp-client@5.13.1" */
  id: string;
  /** "HIGH",... */
  confidence: string;
  url: string;
}

interface IPackageVulnerabilities {
  used: number;
  vulnerabilities: IVulnerability[];
}
