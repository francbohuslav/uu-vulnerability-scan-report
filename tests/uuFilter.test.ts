import { INode, WhiteListTypeEnum } from "../src/interfaces";
import { UuFilter } from "../src/uuFilter";

let nodes: INode[] = [];
beforeEach(() => {
  nodes = [
    {
      library: "org.apache.camel:camel-core",
      childNodes: [
        {
          library: "org.apache.camel:camel-bean",
          childNodes: [],
          level: 2,
          version: "3.11.1",
          vulnerabilities: [],
        },
      ],
      level: 1,
      version: "3.11.1",
      vulnerabilities: [],
    },
    {
      library: "Enelane",
      childNodes: [
        {
          library: "uu.appg01:uu_appg01_server",
          childNodes: [],
          level: 2,
          version: "5.2.1",
          vulnerabilities: [],
        },
      ],
      level: 1,
      version: "5.0.4-SNPSHOT",
      vulnerabilities: [],
    },
    {
      library: "uu.appg01:uu_appg01_auditlog",
      childNodes: [],
      level: 1,
      version: "5.2.1",
      vulnerabilities: [],
    },
  ];
});
describe(`UU filter`, () => {
  test(`No filter`, () => {
    const filter = new UuFilter();
    const result = filter.filter(nodes, WhiteListTypeEnum.all);
    expect(result).toHaveLength(3);
    expect(result[1].childNodes).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  test(`No UU`, () => {
    const filter = new UuFilter();
    const result = filter.filter(nodes, WhiteListTypeEnum.exceptUu);
    expect(result).toHaveLength(2);
    expect(result[1].childNodes).toHaveLength(0);
    expect(result).toMatchSnapshot();
  });

  test(`Only UU`, () => {
    const filter = new UuFilter();
    const result = filter.filter(nodes, WhiteListTypeEnum.onlyUu);
    expect(result).toHaveLength(2);
    for (const node of result) {
      expect(node.library.startsWith("uu.appg01:")).toBeTruthy();
    }
    expect(result).toMatchSnapshot();
  });
});
