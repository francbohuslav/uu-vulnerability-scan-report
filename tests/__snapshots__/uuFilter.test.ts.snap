// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UU filter No UU 1`] = `
[
  {
    "childNodes": [
      {
        "childNodes": [],
        "level": 2,
        "library": "org.apache.camel:camel-bean",
        "version": "3.11.1",
        "vulnerabilities": [],
      },
    ],
    "level": 1,
    "library": "org.apache.camel:camel-core",
    "version": "3.11.1",
    "vulnerabilities": [],
  },
  {
    "childNodes": [],
    "level": 1,
    "library": "Enelane",
    "version": "5.0.4-SNPSHOT",
    "vulnerabilities": [],
  },
]
`;

exports[`UU filter No filter 1`] = `
[
  {
    "childNodes": [
      {
        "childNodes": [],
        "level": 2,
        "library": "org.apache.camel:camel-bean",
        "version": "3.11.1",
        "vulnerabilities": [],
      },
    ],
    "level": 1,
    "library": "org.apache.camel:camel-core",
    "version": "3.11.1",
    "vulnerabilities": [],
  },
  {
    "childNodes": [
      {
        "childNodes": [],
        "level": 2,
        "library": "uu.appg01:uu_appg01_server",
        "version": "5.2.1",
        "vulnerabilities": [],
      },
    ],
    "level": 1,
    "library": "Enelane",
    "version": "5.0.4-SNPSHOT",
    "vulnerabilities": [],
  },
  {
    "childNodes": [],
    "level": 1,
    "library": "uu.appg01:uu_appg01_auditlog",
    "version": "5.2.1",
    "vulnerabilities": [],
  },
]
`;

exports[`UU filter Only UU 1`] = `
[
  {
    "childNodes": [],
    "level": 2,
    "library": "uu.appg01:uu_appg01_server",
    "version": "5.2.1",
    "vulnerabilities": [],
  },
  {
    "childNodes": [],
    "level": 1,
    "library": "uu.appg01:uu_appg01_auditlog",
    "version": "5.2.1",
    "vulnerabilities": [],
  },
]
`;
